django/contrib/admin/widgets.py

```diff
@@ -397,3 +399,91 @@ def __init__(self, attrs=None):
 
 class AdminBigIntegerFieldWidget(AdminIntegerFieldWidget):
     class_name = 'vBigIntegerField'
+
+
+class AutocompleteMixin(object):
+    """
+    ForeignKey widget that loads options via AJAX to avoid rendering large Querysets.
+
+    This class is responsible for rendering the necessary
+    data attributes for select2 as well as adding the static
+    form media.
+    """
+
+    url_name = 'admin:autocomplete'
+
+    def __init__(self, model_admin, attrs=None, using=None):
+        self.model_admin = model_admin
+        self.db = using
+        super(AutocompleteMixin, self).__init__(attrs)
+
+    def get_url(self, name):
+        """Return URL from instance or by reversing :attr:`.url_name`."""
+        return reverse(self.url_name, kwargs={
+            'app_label': self.model_admin.model._meta.app_label,
+            'model_name': self.model_admin.model.__name__,
+            'field_name': self.get_field_name(name),
+        })
+
+    def build_attrs(self, extra_attrs=None, **kwargs):
+        """Set select2's AJAX attributes."""
+        name = kwargs['name']
+        attrs = super(AutocompleteMixin, self).build_attrs(extra_attrs=extra_attrs, **kwargs)
+        attrs.setdefault('class', '')
+        attrs.update({
+            'data-ajax--url': self.get_url(name),
```

Comment:
That is how nested options can be described since `Select2` v4 in a data attribute.
https://select2.github.io/options.html#how-can-i-set-a-default-value-for-a-nested-option